### ISSUE TITLE
Optimize Hero Build Optimization Code

### ISSUE THREAD

**Issue by kenibrewer** [LukeCroteau/RaidExtractor#8]
Currently the hero build optimization code relies on brute forcing every possible combination of artifacts and calculating the stats. This could be sped up considerably by taking some more intelligent approaches to this variation on the subset sum / optimization problem (NP-hard btw).

Luke has some code built from a previous attempt solving this problem that might be possible to simply slot in. I'll let him describe it, but I imagine it's probably written in python though which would be adding a new programming language to the project. That's probably still the best starting point though. 

If I were to tackle this de novo, I'd be inclined to building sorted lists of artifacts by different stat bonuses and using some variation of a local search algorithm to try swapping in different gear pieces and see the swaps increase or decrease the build score.

**Comment by kanadra** [comment]
Pardon the wall of text in advance, but my personal aproach would be to **run premutations and use the provided data to power a smart gearing process**.

> Example:
> A champion needs only 2 speed added across 7 gear slots with speed stats/substats.
> 
> So the premutation (Bruteforcing but this can be turned into a function as well) would find all the possible combinations of speed to get to 2 speed, which would return the following list with a total of 28 combinations (This list should never be longer than 1,012)
> 1, 1, 0, 0, 0, 0, 0
> 1, 0, 1, 0, 0, 0, 0
> 1, 0, 0, 1, 0, 0, 0
> 1, 0, 0, 0, 1, 0, 0
> 1, 0, 0, 0, 0, 1, 0
> 1, 0, 0, 0, 0, 0, 1
> 2, 0, 0, 0, 0, 0, 0
> etc...

This provides an easy **list of instructions to allow searching for specific values** while removing all the gear that cannot be used to achieve the desired result. 0 values are not populated until the next priority stat is calculated.

This should dramatically reduce the time it takes to find all usable combinations, however its **performance improvments in wider/open parameters is less noticable** (Eg. I want my atk stat to be as high as possible).

Name/Abv/Max
Artifacts/A/700
Constraint range/CR/User input dependant
Stat range/SR/1,012
Stat count/SC/8

Big O notation should be reduced from:
O = A^A = **Max of infinite attempts** (Calculators return overflow errors or display infinity)
O = A*CR*SR*SC = **Max of 283,360,000 attempts** (Asuming a constraint range of 50 per stat)

This time could be **further reduced with a desicion tree**, however that seems unnessesary since it is unlikely that anyone can achieve the desired results with more than a few hundred combinations.

**Comment by LukeCroteau** [comment]
I'm going to define this particular item as Out of Scope, unless someone wants to step up and work on this functionality.

I'm already aware of a few other projects looking to take over this space, but would gladly accept submissions.

Leaving this issue open until 2021-01-31.

**Comment by cmarabate** [comment]
Yeah, I think everyone realizes that they are not going to be able to do better than the original developer of RaidExtractor did with HellHades' Raid Optimizer. Best to focus on other stuff we can do with the data extracted. I have never saw much point in trying to develop something that already exists unless I know that I can improve upon it 😁.
